src/librustc_borrowck/borrowck/mir/mod.rs

```diff
@@ -89,3 +192,127 @@ impl<'b, 'a: 'b, 'tcx: 'a> MirBorrowckCtxt<'b, 'a, 'tcx> {
         debug!("MirBorrowckCtxt::process_terminator({:?}, {:?})", bb, term);
     }
 }
+
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+enum DropFlagState {
+    Live,
+    Dead
+}
+
+fn on_all_children_bits<'a, 'tcx, F>(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    mir: &Mir<'tcx>,
+    move_data: &MoveData<'tcx>,
+    move_path_index: MovePathIndex,
+    mut each_child: F)
+    where F: FnMut(MovePathIndex)
+{
+    fn is_terminal_path<'a, 'tcx>(
+        tcx: TyCtxt<'a, 'tcx, 'tcx>,
+        mir: &Mir<'tcx>,
+        move_data: &MoveData<'tcx>,
+        path: MovePathIndex) -> bool
+    {
+        match move_data.move_paths[path].content {
+            MovePathContent::Lvalue(ref lvalue) => {
+                match mir.lvalue_ty(tcx, lvalue).to_ty(tcx).sty {
+                    // don't trace paths past arrays, slices, and
+                    // pointers. They can only be accessed while
+                    // their parents are initialized.
+                    //
+                    // FIXME: we have to do something for moving
+                    // slice patterns.
+                    ty::TyArray(..) | ty::TySlice(..) |
+                    ty::TyRef(..) | ty::TyRawPtr(..) => true,
+                    _ => false
+                }
+            }
+            _ => true
+        }
+    }
+
+    fn on_all_children_bits<'a, 'tcx, F>(
+        tcx: TyCtxt<'a, 'tcx, 'tcx>,
+        mir: &Mir<'tcx>,
+        move_data: &MoveData<'tcx>,
+        move_path_index: MovePathIndex,
+        each_child: &mut F)
+        where F: FnMut(MovePathIndex)
+    {
+        each_child(move_path_index);
+
+        if is_terminal_path(tcx, mir, move_data, move_path_index) {
+            return
+        }
+
+        let mut next_child_index = move_data.move_paths[move_path_index].first_child;
+        while let Some(child_index) = next_child_index {
+            on_all_children_bits(tcx, mir, move_data, child_index, each_child);
+            next_child_index = move_data.move_paths[child_index].next_sibling;
+        }
+    }
+    on_all_children_bits(tcx, mir, move_data, move_path_index, &mut each_child);
+}
+
+fn drop_flag_effects_for_function_entry<'a, 'tcx, F>(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    mir: &Mir<'tcx>,
+    move_data: &MoveData<'tcx>,
+    mut callback: F)
+    where F: FnMut(MovePathIndex, DropFlagState)
+{
+    for i in 0..(mir.arg_decls.len() as u32) {
+        let lvalue = repr::Lvalue::Arg(i);
+        let move_path_index = move_data.rev_lookup.find(&lvalue);
+        on_all_children_bits(tcx, mir, move_data,
+                             move_path_index,
+                             |moi| callback(moi, DropFlagState::Live));
+    }
```

Comment:
Ah, so @pnkfelix poined me at this line:

https://github.com/rust-lang/rust/pull/33667/files#diff-1246b8e41df77c9faa592a15189f9b2fR954

which I had overlooked. Seems good!